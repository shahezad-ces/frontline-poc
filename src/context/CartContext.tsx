"use client";
import { createContext, useContext, useEffect, useState } from "react";
import { getLocalStorage, setLocalStorage } from "@frontline/utils";
import type { CartContextType, CartItem, Product } from "@frontline/types";

const CartContext = createContext<CartContextType | undefined>(undefined);

export const CartProvider = ({ children }: { children: React.ReactNode }) => {
  const [cart, setCart] = useState<CartItem[]>([]);
  const [quantity, setQuantity] = useState(1);

  // Load from localStorage
  useEffect(() => {
    const storedCart = getLocalStorage<CartItem[]>("cart");
    if (storedCart) setCart(storedCart);
  }, []);

  // Save to localStorage
  useEffect(() => {
    setLocalStorage("cart", cart);
  }, [cart]);

  const addToCart = (product: Product) => {
    setCart((prev) => {
      const existing = prev.find((item) => item.id === product.id);
      if (existing) {
        return prev.map((item) =>
          item.id === product.id
            ? { ...item, quantity: item.quantity + quantity }
            : item
        );
      }
      return [...prev, { ...product, quantity }];
    });
    setQuantity(1);
  };

  const updateQuantity = (quantity: number) => {
    if (quantity < 1) return; // Prevent zero or negative
    setQuantity(quantity);
  };

  const removeFromCart = (id: number) => {
    setCart((prev) => prev.filter((item) => item.id !== id));
  };

  const clearCart = () => setCart([]);

  const totalItems = cart.reduce((sum, item) => sum + item.quantity, 0);
  const totalPrice = cart.reduce(
    (sum, item) => sum + item.price * item.quantity,
    0
  );

  return (
    <CartContext.Provider
      value={{
        cart,
        addToCart,
        updateQuantity,
        removeFromCart,
        clearCart,
        totalItems,
        totalPrice,
        quantity,
      }}
    >
      {children}
    </CartContext.Provider>
  );
};

export const useCart = () => {
  const context = useContext(CartContext);
  if (!context) throw new Error("useCart must be used within CartProvider");
  return context;
};
