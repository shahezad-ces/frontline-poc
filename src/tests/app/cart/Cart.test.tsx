import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import CartPage from "@frontline/app/cart/page";
import { useCart } from "@frontline/context/CartContext";

// Mock useCart hook
jest.mock("@frontline/context/CartContext", () => ({
  useCart: jest.fn(),
}));

// Mock child components
jest.mock("@frontline/stories", () => ({
  Image: ({ src, alt }: any) => (
    <img src={src} alt={alt} data-testid="cart-image" />
  ),
  QuantitySelector: () => (
    <div data-testid="quantity-selector">Quantity Selector</div>
  ),
  Button: ({ children, ...props }: any) => (
    <button {...props}>{children}</button>
  ),
}));

jest.mock("next/link", () => ({
  __esModule: true,
  default: ({ children, href }: any) => <a href={href}>{children}</a>,
}));

const mockUseCart = useCart as jest.Mock;

const mockCartItem = {
  id: 1,
  title: "Test Product",
  price: 100,
  images: ["https://placehold.co/100x100"],
  category: { id: 1, name: "Electronics", image: "cat.jpg" },
  quantity: 2,
};

describe("CartPage Component", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders empty cart message when cart is empty", () => {
    mockUseCart.mockReturnValue({
      cart: [],
      totalPrice: 0,
      removeFromCart: jest.fn(),
    });

    render(<CartPage />);

    expect(screen.getByText(/Your cart is empty/i)).toBeInTheDocument();
    expect(
      screen.getByText(/Add products you want to buy later here/i)
    ).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: /See Offers/i })
    ).toBeInTheDocument();
  });

  it("renders cart items when cart has products", () => {
    mockUseCart.mockReturnValue({
      cart: [mockCartItem],
      totalPrice: 200,
      removeFromCart: jest.fn(),
    });

    render(<CartPage />);

    expect(screen.getByText(/My Cart/i)).toBeInTheDocument();
    expect(screen.getByText(/Test Product/i)).toBeInTheDocument();
    expect(screen.getByText(/Electronics/i)).toBeInTheDocument();
    expect(screen.getByText(/\$100/i)).toBeInTheDocument();
    expect(screen.getByTestId("cart-image")).toHaveAttribute(
      "src",
      mockCartItem.images[0]
    );
    expect(screen.getByTestId("quantity-selector")).toBeInTheDocument();
  });

  it("calls removeFromCart when Remove button is clicked", () => {
    const removeMock = jest.fn();
    mockUseCart.mockReturnValue({
      cart: [mockCartItem],
      totalPrice: 200,
      removeFromCart: removeMock,
    });

    render(<CartPage />);

    const removeButton = screen.getByRole("button", { name: /Remove/i });
    fireEvent.click(removeButton);

    expect(removeMock).toHaveBeenCalledWith(mockCartItem.id);
  });

  it("Continue Shopping button is disabled when cart is empty", () => {
    mockUseCart.mockReturnValue({
      cart: [],
      totalPrice: 0,
      removeFromCart: jest.fn(),
    });

    render(<CartPage />);
    const continueButton = screen.getByRole("button", { name: /See Offers/i });
    expect(continueButton).toBeEnabled(); // "See Offers" button is always enabled
  });

  it("matches snapshot", () => {
    mockUseCart.mockReturnValue({
      cart: [mockCartItem],
      totalPrice: 200,
      removeFromCart: jest.fn(),
    });

    const { asFragment } = render(<CartPage />);
    expect(asFragment()).toMatchSnapshot();
  });
});
