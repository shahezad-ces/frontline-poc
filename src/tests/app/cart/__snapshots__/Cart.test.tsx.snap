// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CartPage Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto px-4 my-6 sm:px-6 lg:px-8 gap-6"
  >
    <h1
      class="text-3xl mb-4 font-bold text-gray-800"
    >
      My Cart
    </h1>
    <div
      class="flex flex-col lg:flex-row gap-2 md:gap-6"
    >
      <div
        class="w-full lg:w-[65%] flex flex-col gap-2 md:gap-6 bg-white border border-gray-200 rounded-lg p-2 md:p-4"
      >
        <a
          href="/product/1"
        >
          <div
            class="flex items-start flex-col md:flex-row gap-2 md:gap-6 w-full"
          >
            <img
              alt="Test Product"
              data-testid="cart-image"
              src="https://placehold.co/100x100"
            />
            <div
              class="space-y-1"
            >
              <h2
                class="text-l font-semibold text-gray-800 line-clamp-2"
              >
                Test Product
              </h2>
              <p
                class="text-sm text-gray-400"
              >
                Electronics
              </p>
              <p
                class="text-xl text-red-600 font-semibold"
              >
                $100
              </p>
            </div>
          </div>
          <div
            class="flex flex-col self-center gap-2 md:gap-4"
          >
            <div
              data-testid="quantity-selector"
            >
              Quantity Selector
            </div>
            <button
              class="rounded-lg"
              variant="danger"
            >
              <div
                class="flex gap-1 item-center justify-center"
              >
                <svg
                  aria-hidden="true"
                  class="w-6 h-6"
                  data-slot="icon"
                  fill="currentColor"
                  viewBox="0 0 20 20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M8.75 1A2.75 2.75 0 0 0 6 3.75v.443c-.795.077-1.584.176-2.365.298a.75.75 0 1 0 .23 1.482l.149-.022.841 10.518A2.75 2.75 0 0 0 7.596 19h4.807a2.75 2.75 0 0 0 2.742-2.53l.841-10.52.149.023a.75.75 0 0 0 .23-1.482A41.03 41.03 0 0 0 14 4.193V3.75A2.75 2.75 0 0 0 11.25 1h-2.5ZM10 4c.84 0 1.673.025 2.5.075V3.75c0-.69-.56-1.25-1.25-1.25h-2.5c-.69 0-1.25.56-1.25 1.25v.325C8.327 4.025 9.16 4 10 4ZM8.58 7.72a.75.75 0 0 0-1.5.06l.3 7.5a.75.75 0 1 0 1.5-.06l-.3-7.5Zm4.34.06a.75.75 0 1 0-1.5-.06l-.3 7.5a.75.75 0 1 0 1.5.06l.3-7.5Z"
                    fill-rule="evenodd"
                  />
                </svg>
                Remove
              </div>
            </button>
          </div>
        </a>
      </div>
      <div
        class="w-full lg:w-[35%] space-y-4 bg-white border border-gray-200 rounded-lg p-3 md:p-4 h-fit"
      >
        <h2
          class="text-l md:text-xl font-semibold text-gray-800"
        >
          Summary of your purchase
        </h2>
        <div
          class="border border-gray-200 rounded-lg p-2 space-y-1"
        >
          <div
            class="flex justify-between"
          >
            <p>
              Subtotal (1 products)
            </p>
            <p>
              $200
            </p>
          </div>
          <div
            class="flex justify-between"
          >
            <p>
              Discount
            </p>
            <p>
              $0
            </p>
          </div>
          <div
            class="flex justify-between"
          >
            <p>
              Shipment
            </p>
            <p>
              Free
            </p>
          </div>
        </div>
        <div
          class="bg-neutral-100 flex justify-between items-center border border-gray-200 rounded-lg p-2 space-y-1 mt-1"
        >
          <p
            class="font-semibold"
          >
            Total cash (VAT included)
          </p>
          <p
            class="font-semibold"
          >
            $200
          </p>
        </div>
        <button
          class="w-full"
        >
          Continue Shopping
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
